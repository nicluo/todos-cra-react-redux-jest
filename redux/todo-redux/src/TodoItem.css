.Checkbox {
    width: 30px;
    height: 30px;
    margin-right: 30px;
    background-color: white;
    border-radius: 50%;
    border: 1px solid #ddd;
    -webkit-appearance: none;
    outline: none;
    cursor: pointer;
}

.Checkbox:checked {
    background-image: url('Check.svg');
    background-size: 30px;
}
