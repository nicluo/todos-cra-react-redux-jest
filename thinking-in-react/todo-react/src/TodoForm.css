.Input {
    background: none;
    outline: none;
    border: none;
    border-bottom: 1px solid #ddd;
    padding: 4px;
    flex-grow: 1;
}

.SubmitButton {
    background: none;
    border: none;
    outline: none;
    width: 30px;
    height: 30px;
    margin-right: 30px;
    border-radius: 50%;
    cursor: pointer;
}

.SubmitButton:hover {
    background: #ddd;
}

.Symbol {
    width: 30px;
}
