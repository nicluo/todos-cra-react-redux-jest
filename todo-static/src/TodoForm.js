import React, { Component } from 'react';
import PropTypes from 'prop-types';

class TodoForm extends Component {
  render() {
    const {input, onChange, onSubmit} = this.props;

    return (
      <div className="TodoForm center mw-100 mw8-ns ba bw1 b--black-40 br2 mv3">
        <form onSubmit={onSubmit}>
          <div className="flex items-center pa3 f4 f4-ns lh-copy">
            <button type="submit" className="SubmitButton f3 dt">
              <div className="Symbol dtc">+</div>
            </button>
            <input type="text" className="Input mw-100" value={input} onChange={onChange} />
          </div>
        </form>
      </div>
    );
  }
}

TodoForm.propTypes = {
  input: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

export default TodoForm;
