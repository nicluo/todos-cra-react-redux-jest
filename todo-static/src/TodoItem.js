import React, { Component } from 'react';
import PropTypes from 'prop-types';

class TodoItem extends Component {
  handleChange = (event) => {
    this.props.onChange(this.props.id, event.target.checked);
  };

  render() {
    const {completed, children} = this.props;

    return (
      <div className="TodoItem center mw-100 mw8-ns ba bw1 b--black-40 br2 mv3">
        <div className="flex items-center pa3 f4 f4-ns lh-copy">
          <input type="checkbox" className="Checkbox" checked={completed} onChange={this.handleChange} />
          <div className="dib">{children}</div>
        </div>
      </div>
    );
  }
}

TodoItem.propTypes = {
  id: PropTypes.number.isRequired,
  completed: PropTypes.bool.isRequired,
  children: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
};

export default TodoItem;
