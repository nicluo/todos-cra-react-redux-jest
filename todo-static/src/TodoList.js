import React, { Component } from 'react';
import TodoItem from "./TodoItem";
import TodoForm from "./TodoForm";

const TODOS = [
  {title: "Start With A Mock", completed: true},
  {title: "Step 1: Break The UI Into A Component Hierarchy", completed: true},
  {title: "Step 2: Build A Static Version in React", completed: true},
  {title: "Step 3: Identify The Minimal (but complete) Representation Of UI State", completed: false},
];

const TODOFORM = {
  input: "New Todo Item"
};

class TodoList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      todos: TODOS,
      todoForm: TODOFORM
    };
  }

  render() {
    const {todos, todoForm} = this.state;

    const todoListContent = todos.map(function(todo) {
      return (<TodoItem completed={todo.completed}>{todo.title}</TodoItem>);
    });

    return (
      <div className="TodoList ph3 mv4">
        {todoListContent}
        <TodoForm {...todoForm} />
      </div>
    );
  }
}

export default TodoList;
