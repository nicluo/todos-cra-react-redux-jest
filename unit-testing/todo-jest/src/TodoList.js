import React, { Component } from 'react';
import TodoItem from "./TodoItem";
import TodoForm from "./TodoForm";

const DEFAULT_STATE = {
  todos: [],
  input: ""
};

class TodoList extends Component {
  state = DEFAULT_STATE;

  handleInputChange = (event) => {
    this.setState({input: event.target.value});
  };

  handleSubmit = (event) => {
    this.setState((prevState) => {
      const newTodos = prevState.todos.slice();
      const todo = {
        // TODO: Generate better unique ids rather than using Math.random()
        id: Math.floor(Math.random() * 1000),
        title: prevState.input,
        completed: false
      };
      newTodos.push(todo);
      return {todos: newTodos, input: ''};
    });
    event.preventDefault();
  };

  handleTodoChange = (id, completed) => {
    this.setState((prevState) => {
      const newTodos = this.state.todos.map((todo) => {
        if (todo.id === id) {
          todo.completed = completed
        }
        return todo;
      });
      return {todos: newTodos};
    });
  };

  render() {
    const {todos, input} = this.state;

    const todoListContent = todos.map((todo) => {
      return (<TodoItem key={todo.id} id={todo.id} completed={todo.completed} onChange={this.handleTodoChange}>{todo.title}</TodoItem>);
    });

    return (
      <div className="TodoList ph3 mv4">
        {todoListContent}
        <TodoForm onChange={this.handleInputChange} onSubmit={this.handleSubmit} input={input} />
      </div>
    );
  }
}

export default TodoList;
