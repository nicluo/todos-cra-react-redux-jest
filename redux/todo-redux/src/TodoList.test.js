import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import TodoList from './TodoList';
import TodoForm from './TodoForm';
import TodoItem from './TodoItem';

it('Renders initial state without crashing', () => {
  const component = shallow(
    <TodoList />
  );
  expect(shallowToJson(component)).toMatchSnapshot();
});

it('Should have a TodoForm item', () => {
  const component = shallow(
    <TodoList />
  );
  expect(component.find(TodoForm).length).toEqual(1);
});

it('Should update state input when TodoForm is changed', () => {
  let value = 'test value';
  const component = shallow(
    <TodoList />
  );
  component.find(TodoForm).simulate('change', {target: {value: value}});
  expect(component.state('input')).toEqual(value);
});

it('Should create a new TodoItem when handleSubmit is called', () => {
  const component = shallow(
    <TodoList />
  );
  component.find(TodoForm).simulate('submit', {preventDefault: () => {}});
  expect(component.find(TodoItem).length).toEqual(1);
});

it('Should update state todos in state when handleTodoChange is called', () => {
  let state = {
    todos: [
      {id: 1, title: 'test item', completed: false},
      {id: 2, title: 'test item', completed: false}
      ],
    input: ''
  };
  const component = shallow(
    <TodoList />
  );
  component.setState(state);
  component.find(TodoItem).first().simulate('change', 1, true);
  expect(component.state('todos')[0].completed).toEqual(true);
  expect(component.state('todos')[1].completed).toEqual(false);
});
