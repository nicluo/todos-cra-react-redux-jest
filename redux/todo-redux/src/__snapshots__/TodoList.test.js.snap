// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders initial state without crashing 1`] = `
<div
  className="TodoList ph3 mv4"
>
  <TodoForm
    input=""
    onChange={[Function]}
    onSubmit={[Function]}
  />
</div>
`;
